src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -148,5 +162,94 @@ public virtual void NextBytes(byte[] buffer)
 
         private static void ThrowMinMaxValueSwapped() =>
             throw new ArgumentOutOfRangeException("minValue", SR.Format(SR.Argument_MinMaxValue, "minValue", "maxValue"));
+
+        /// <summary>Random implementation that delegates all calls to a ThreadStatic Impl instance.</summary>
+        private sealed class ThreadSafeRandom : Random
+        {
+            // We need Random.Shared to return an instance that is thread-safe, as it may be used from multiple threads.
+            // It's also possible that one thread could retrieve Shared and pass it to another thread, so Shared can't
+            // just access a ThreadStatic to return a Random instance stored there.  As such, we need the instance
+            // returned from Shared itself to be thread-safe, which can be accomplished either by locking around all
+            // accesses or by itself accessing a ThreadStatic on every access: we've chosen the latter, as it is more
+            // scalable.  With that, we have two basic approaches:
+            // 1. the instance returned can be a base Random instance constructed with an _impl that is a ThreadSafeImpl.
+            // 2. the instance returned can be a special Random-derived instance, where _impl is ignored and the derived
+            //    type overrides all methods on the base.
+            // (1) is cleaner, as (2) requires duplicating a bit more code, but (2) enables all virtual dispatch to be
+            // devirtualized and potentially inlined, so that Random.Shared.NextXx(...) ends up being faster.
+            // This implements (2).
+
+            [ThreadStatic]
+            private static XoshiroImpl? t_random;
+
+            public ThreadSafeRandom() : base(isThreadSafeRandom: true) { }
+
+            private static XoshiroImpl LocalRandom => t_random ?? Create();
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            private static XoshiroImpl Create() => t_random = new();
```

Comment:
After the first seeding; should this then use the Xoshiro jump functions to generate each new random instance; rather than starting with a fresh random seed?

e.g. from [xoshiro256starstar.c](https://prng.di.unimi.it/xoshiro256starstar.c)
```c++
/* This is the jump function for the generator. It is equivalent
   to 2^128 calls to next(); it can be used to generate 2^128
   non-overlapping subsequences for parallel computations. */
```

Reply:
To what end? Just making first use on a secondary thread faster? How would that thread safely use the original instance?

Reply:
> To what end?

At the moment it is using a secondary random generator `Interop.GetRandomBytes` in `new()` to seed the new sequence; which is fine(ish); however there will be N independent random sequences (i.e. thread count).

Since they are uncoordinated random sequences there is the question of whether that increases the likelyhood of [birthday paradox](https://en.wikipedia.org/wiki/Birthday_problem) repeats and shortening the period of the randomness when used in parallel.

However the algorithms do include jump functions specifically to provide the next seed for concurrent use; so it might be better to do that so the seeds are coordinated?

> How would that thread safely use the original instance?

Same as now; just lock on the creation to jump the seed in a coordinated fashion, is just the create of a new `ThreadStatic` instance that would benefit from coordination, the usage wouldn't change or require any locking (i.e. no changes to the rest of the implementation)

Reply:
Testing the first `ushort.MaxValue` `ulong`s of `ushort.MaxValue` Randoms on .NET 6 for collisions; which is a 32bit sample of a 64bit range; so shouldn't have collisions:
```csharp
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

var values = new HashSet<ulong>();

for (int i = 0; i < ushort.MaxValue; i++)
{
    var rand = new Random();
    for (var c = 0; c < ushort.MaxValue; c++)
    {
        var value = 0ul;
        rand.NextBytes(MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref value, 1)));

        if (!values.Add(value))
        {
            Console.WriteLine($"Collision on Random {i} at value {c}");
        }
    }
}

Console.WriteLine("Done");
```
I run out of memory before getting a collision; so maybe it doesn't matter 😅

However it does entirely depend on the behaviour of `Interop.GetRandomBytes` as the seed; rather than being internally consistent (which the jump function should provide).

Reply:
Why would this concern be relevant to the instances created for Random.Shared but not to `new Random()` in general?

From my perspective, this is not something to be concerned about.  I'm open to other opinions, though.

Reply:
The nature of `.Shared` is its able to work across multiple threads; and has implication that it's distribution as good as `Random` used on a single thread.

However given the test https://github.com/dotnet/runtime/pull/50297#discussion_r602740716 didn't show up any collisions; its probably not worth worrying about.